moirae/estimators/online/utils/model.py

```diff
@@ -0,0 +1,153 @@
+#  the `predict_output` function first estimates how the transients will update for each set of ASOH,
+#  then uses the updated transient states and ASOH to determine the outptus
+
+class CellModelInterface(ModelWrapper):
+    """Link between the :class:`~moirae.model.base.CellModel` and the numpy-only interface of
+    the filter implementations."""
+
+    model: CellModel
+    """Cell model underpinning the update functions"""
+    asoh: HealthVariable
+    """ASOH values passed to each call of the cell model"""
+    transients: GeneralContainer
+    """Transient states used for the inputs of the model"""
+
+    def __init__(self,
+                 cell_model: CellModel,
+                 asoh: HealthVariable,
+                 transients: GeneralContainer,
+                 input_template: InputQuantities):
+
+        # Store the ASOH and transient state, making sure they are not batched
+        if asoh.batch_size > 1:
+            raise ValueError(f'The batch size of the ASOH must be 1. Found: {asoh.batch_size}')
+        if transients.batch_size > 1:
+            raise ValueError(f'The batch size of the transient state must be 1. Found: {transients.batch_size}')
+
+        self.transients = transients
+        self.model = cell_model
+        self.asoh = asoh
+        self.input_template = input_template
+
+        # Capture the shape of the outputs
+        self._num_output_dimensions = self.model.calculate_terminal_voltage(self.input_template,
+                                                                            self.transients,
+                                                                            self.asoh).to_numpy().shape[1]
+
+    @property
+    def num_output_dimensions(self) -> int:
+        return self._num_output_dimensions
+
+
+class JointCellModelInterface(CellModelInterface):
+    """Interface used when the hidden state used by a filter includes the transient states.
+
+    Create the interface by defining
+        - Which portions of the ASOH are used as inputs to function
+        - Values for the ASOH parameters that remain fixed
+        - An example transient state and input to be passed to the function which will be used as a template
+
+    The resultant function will take numpy arrays as inputs and produce numpy arrays as outputs
+
+    Args:
+        model: Model which defines the physics of the system being modeled
+        asoh: Values for all state of health parameters of the model
+        transients: Current values of the transient state of the system
+        input_template: Example input values for the model
+        asoh_inputs: Names of the ASOH parameters to include as part of the hidden state
+    """
+
+    def __init__(self,
+                 model: CellModel,
+                 asoh: HealthVariable,
+                 transients: GeneralContainer,
+                 input_template: InputQuantities,
+                 asoh_inputs: Tuple[str]):

```

Comment:
Would it be possible to make this optional, and, when not provided, take everything that is updatable?

Reply:
Totally. I just left the defaults off since I didn't know what sensible values would be

Reply:
i.e., and opted for being super-specific